### ISSUE TITLE
Usage of available_processor_count in default puma config leads to excessive number of puma workers in hosting platforms with shared vCPU management and containerized environment

### ISSUE THREAD

**Issue by denys-chaikovskyi** [rails/rails#52522]
### Description

The default Puma configuration in Rails 7.1.3.4 leverages the `Concurrent.physical_processor_count` to automatically detect a number of Puma workers that can be spawned by default based on available CPUs.

I see that it was later removed ([commit](https://github.com/rails/rails/commit/06d614ada9e4609ff83659e842f48af3232a03a5)) and recently `Concurrent.available_processor_count` was added instead ([commit](https://github.com/rails/rails/commit/f719787c582839fd2fcd886d70b43da3ddad2ceb)), which is a group-aware method to detect how many processors we can actually use on various shared hosting platforms.

I tried using both `Concurrent.physical_processor_count` and `Concurrent.available_processor_count` on an instance of Digital Ocean App Platform with 1 shared vCPU and 2GB RAM. Unfortunately, both methods failed to identify a real count of available CPUs spawning too many workers, which leads to excessive RAM usage (70-95% with minimal user activity) and degraded performance.

### Steps to reproduce
1. Deploy a Rails application on the Digital Ocean App Platform with 1 shared vCPU and 2GB RAM connected to the managed database. (The "App Platform" is a Paas from DO that is similar to Heroku) 
![CleanShot 2024-08-06 at 19 48 15@2x](https://github.com/user-attachments/assets/09720fab-3592-496b-9b23-87077e325db4)

2. Use the default `puma.rb` configuration provided by Rails 7.2.0.rc1 (which uses  `Concurrent.available_processor_count` method)

3. Open Digital Ocean App Platform instance's console and check the number of spawned workers with a command like `ps aux --sort=-%mem | head -n 10` 

4. Launch the Rails console within the same console session and check what value `Concurrent.available_processor_count` returns.

### Expected behavior
For Digital Ocean App Platform instance with 1 shared vCPU, puma.rb should not spawn multiple puma workers. `workers_count` should be 1.

In my experience of running a similar fresh new MVP (minimal user usage) with similar setup, RAM usage should be in the range of 15-25%.

### Actual behavior
The number of Puma workers is set incorrectly for shared virtual CPU, spawning too many workers, which leads to excessive RAM usage (70-95% with minimal user activity) and degraded performance.

In my case I got 8 puma workers:
```bash
rails@appname:/rails$ ps aux --sort=-%mem | head -n 10
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
rails       69  0.0 15.3 905500 321880 ?       Sl   Aug05   0:53 puma: cluster worker 6: 1 [rails]
rails       35  0.0 15.1 896668 317424 ?       Sl   Aug05   0:47 puma: cluster worker 0: 1 [rails]
rails       51  0.0 15.1 901404 317372 ?       Sl   Aug05   0:41 puma: cluster worker 4: 1 [rails]
rails       38  0.0 15.1 902428 317012 ?       Sl   Aug05   0:56 puma: cluster worker 1: 1 [rails]
rails       76  0.0 14.8 895752 312028 ?       Sl   Aug05   0:20 puma: cluster worker 7: 1 [rails]
rails       45  0.0 14.8 894024 311368 ?       Sl   Aug05   0:21 puma: cluster worker 3: 1 [rails]
rails       61  0.0 14.7 896604 310340 ?       Sl   Aug05   0:53 puma: cluster worker 5: 1 [rails]
rails       40  0.0 13.9 876552 292144 ?       Sl   Aug05   0:19 puma: cluster worker 2: 1 [rails]
rails        1  0.0 10.4 808944 218888 ?       Ssl  Aug05   0:34 puma 6.4.2 (tcp://0.0.0.0:3000) [rails]
```
By running Rails console within an instance I can see that both `available_processor_count` and `physical_processor_count` returns number of available CPUs as `8`

```
irb(main):001> Concurrent.available_processor_count
=> 8.0

irb(main):002> Concurrent.physical_processor_count
=> 8
```

From my reading of DO docs, the instance described above doesn't provide a real 8 cores. It most likely shares them among many customers. 

### System configuration
**Ruby version**: 3.3.4
**Rails Version**: 7.2.0.rc1 (experiencing same issue in 7.1.3.4)
**Puma Version**: 6.4.2
**concurrent-ruby version**: 1.3.3
**Hosting Platform**: DigitalOcean App Platform (Docker based)
**Environment**: Production

### Additional information

I'm not sure which mechanism DO App Platform utilizes to manage shared virtual CPUs. 

My assumption is that either accurate cgroud quota information is not available for reading within their Docker container or they use another mechanism for managing shared CPUs across customers. 

For cgroup quota, my instance shows the following:
```
$ cat /sys/fs/cgroup/cpu/cpu.cfs_quota_us
-1
```
-1, from my reading, means no limit on CPU usage. Considering this is a value within the Docker container, it doesn’t necessarily mean there are no limits. My assumption is that a similar issue can occur in other Paas platforms with containerized/virtualized environments.

I think it's a great idea to automatically set the number of Puma workers based on available CPU cores. However, this can be problematic in containerized environments and some PaaS instances with shared CPUs.

If my understanding of the this issue is correct, my concern is that the current default puma.rb configuration might overwhelm new users deploying their first Rails apps on low-resource instances without prior experience and expectations about how much resources a simple Rails app is supposed to consume. This could discourage them from working with a wonderful Rails ecosystem.

Although, we can partially mitigate the issue by also considering available RAM by using something like gem 'sys-memory' when deciding how many puma workers to start. Reading the number of available CPU cores with `Concurrent.available_processor_count` still won't be very accurate on certain platforms with virtualized environments where cgroup quota is not available for reading within the container or when an alternative CPU management mechanism is used. 

cc: @byroot, @dhh 

## P.S.

If I missed or misunderstood something, I would appreciate any clarification.

Thanks!

**Comment by dhh** [comment]
Thanks for the very diligent investigation, @denys-chaikovskyi. I fully agree that we should try hard to ensure that the out-of-the-box experience on a place like DO is solid. I've been using DO for a lot of testing, but hadn't noticed this problem when using Droplets.

I think ultimately it's going to be hard/impossible to find something that works in all instances. But maybe we can build up some combined heuristics that can help guide us. And maybe we can elevate the visibility of the WORKER count/variable. Perhaps using the new default config/deploy.yml file? Since someone would have to modify that anyway to do a deploy with the new Rails defaults, so they're likely to be exposed to it there. Versus normally you don't have to change the puma.rb file.

**Comment by rafaelfranca** [comment]
I think we discussed this problem in the original issue and we decide that we can't possibly support the platform that wrongly report the CPU counts and that is why we kept `WEB_CONCURRENCY`.

I think we should just report the worker count and I agree adding `WEB_CONCURRENCY` to `cofing/deploy.yml` and documenting this problem would be enough.

**Comment by dhh** [comment]
Yeah, I think actually maybe that's the way to just do it. Set WEB_CONCURRENCY (crikey I hate that name) in config/deploy.yml with a comment that you can remove it to get "available processor count". And we can just set it conservatively to 1. That means if you do nothing, you get 1 worker. But you'll be exposed to that fact as you set the IP of your deploy target. And if you remove the setting, you'll get available processor count. Please do explore a PR, if you fancy, @denys-chaikovskyi.

**Comment by rafaelfranca** [comment]
We can also rename `WEB_CONCURRENCY`. We own this variable.

**Comment by byroot** [comment]
Yeah, unfortunately we can't trust that all providers will expose limits in a standard way, and I was definitely expecting it wouldn't work everywhere, but DO is big enough that it kinda kills the idea.

I'm very surprised by that `-1` quota though, I can't image DO went out of their way to mask this value, so I suspect they're not actually limiting CPU usage.

But even then, it's not uncommon not to enforce CPU quota and let the scheduler ensure fairness, so I guess it's not that weird.

**Comment by denys-chaikovskyi** [comment]
Thanks for your prompt replies @dhh, @rafaelfranca and @byroot 

Droplets might not have this problem. I haven't tried using `available_processor_count` there. It might be that DO uses different mechanisms of managing shared CPUs in Droplet vs their PaaS "App Platform". I will have to play with Droplet to clarify that.

Good idea on explicitly setting Worker count/variable in the deploy.yml file to elevate the visibility. I think this is a great improvement for an out-of-the-box experience. I will explore creating PR for this during the weekend.

While this will improve the experience for new Rails deployments using deploy.yml and Kamal, it won’t address the issue for PaaS platforms like the DO App Platform that use a bare Dockerfile. I understand that it’s challenging to ensure a great out-of-the-box experience for every environment and that deploy.yml approach should be prioritized as the new standard.

In light of this, do you think we should still keep `available_processor_count` as the default in puma.rb? Or would setting it conservatively to 1 in in puma.rb be a better default approach?

So basically, I see two options here:

1. Add Worker count/variable in the deploy.yml with the default as 1. Use it in puma.rb and if it is absent go with `available_processor_count`.
2. Add Worker count/variable in the deploy.yml with the default as 1. Use it in puma.rb and if it is absent go with old default as 1. Meaning we don't rely on `available_processor_count`

I think 2nd option would be a bit more more friendly for Paas with shared CPUs that cannot use deploy.yml. But option 1 is clearly a way more optimal for platforms that report the CPU counts accurately. 

A classic crossroads with trade-offs to consider.

@dhh, @rafaelfranca, and @byroot, which option do you think will be better? Do you have an alternative one?

**Comment by byroot** [comment]
I think we could have `WEB_CONCURRENCY=auto` (or whatever the new name is), so that if you are using a platform that supports it you don't need to repeat yourself. But yes, `1` is a safer default.

**Comment by byroot** [comment]
I don't think https://github.com/rails/rails/commit/e3ec553f8c51fc646cecd7fc5efc7fae61e64827 is enough, because that only concerns Kamal.

**Comment by denys-chaikovskyi** [comment]
I think `WEB_CONCURRENCY=auto` (or another name) suggested by @byroot could be a great middle-ground solution to provide a stable experience for platforms with shared vCPU with or without Kamal

In `deploy.yml` we can mention "auto" as a possible option to use with the caveat-comment written by @dhh

```yml
clear:
    # Set this to the number of cores you wish the application to use on each server.
    # You can use "auto" and the app will attempt to use all available cores (but may guess wrong on some cloud hosts!)
    WEB_CONCURRENCY: 1
```

And in `puma.rb` have something like this:
```ruby
  workers_count = if ENV["WEB_CONCURRENCY"] == "auto"
                    Concurrent.available_processor_count
                  else
                    Integer(ENV.fetch("WEB_CONCURRENCY", 1))
                  end
  workers workers_count if workers_count > 1
  ```
  
  What do you think about this approach @dhh @byroot @rafaelfranca ?

**Comment by rafaelfranca** [comment]
That works for me

**Comment by denys-chaikovskyi** [comment]
Great!

@rafaelfranca, @byroot, @dhh 

I’ve created a pull request https://github.com/rails/rails/pull/52533 to address this issue by following approach outlined above. Also updated docs.  

When you have a chance, please review it and provide any feedback.

Thank you!
